### PR TITLE
Update GeoDataFrame to Use the Structured GatherMap Class

### LINKED ISSUE #1222: Fix cuspatial `_gather` API calls to reflect upstream cudf changes
rapidsai/cudf#13534 adds a structured class for gather related methods.
Since cuspatial largely inherits cudf class and rewrites the internal
APIs, cuspatial needs to update the APIs to use the same structures.

### PR DESCRIPTION
**Pull request by isVoid**
## Description

As title, addresses upstream cudf change rapidsai/cudf#13534.

Fixes #1222 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cuspatial/cuspatial/core/geodataframe.py [hunk 1]
```diff
@@ -181,31 +182,29 @@ def _slice(self: T, arg: slice) -> T:
         )
         return self.__class__(result)
 
-    def _apply_boolean_mask(self, mask) -> T:
+    def _apply_boolean_mask(self, gather_map: GatherMap) -> T:
         geo_columns, data_columns = self._split_out_geometry_columns()
-        data = data_columns._apply_boolean_mask(mask)
+        data = data_columns._apply_boolean_mask(gather_map)
```
Comment (wence-) [inline python/cuspatial/cuspatial/core/geodataframe.py:187]: This one is not a `GatherMap` but a [`copy_types.BooleanMask`](https://github.com/rapidsai/cudf/blob/45763fa1f90c174ff9123b4839c729e6c0dee700/python/cudf/cudf/core/copy_types.py#L110).

**Comment by wence-** [comment]
geoseries also needs:
```diff
diff --git a/python/cuspatial/cuspatial/core/geoseries.py b/python/cuspatial/cuspatial/core/geoseries.py
index 2a66ccec..37c75783 100644
--- a/python/cuspatial/cuspatial/core/geoseries.py
+++ b/python/cuspatial/cuspatial/core/geoseries.py
@@ -22,6 +22,7 @@ from shapely.geometry import (
 import cudf
 from cudf._typing import ColumnLike
 from cudf.core.column.column import as_column
+from cudf.core.copy_types import GatherMap
 
 import cuspatial.io.pygeoarrow as pygeoarrow
 from cuspatial.core._column.geocolumn import ColumnType, GeoColumn
@@ -922,10 +923,10 @@ class GeoSeries(cudf.Series):
             aligned_right,
         )
 
-    def _gather(
-        self, gather_map, keep_index=True, nullify=False, check_bounds=True
-    ):
-        return self.iloc[gather_map]
+    def _gather(self, gather_map: GatherMap, keep_index=True):
+        # TODO: This could use the information to avoid reprocessing
+        # in iloc
+        return self.iloc[gather_map.column]
 
     # def reset_index(self, drop=False, inplace=False, name=None):
     def reset_index(
```

**Review by wence- — CHANGES_REQUESTED** [review]
Theres a booleanmask type too!

File: python/cuspatial/cuspatial/core/geodataframe.py (hunk 1 above)
Comment (wence-) [reply python/cuspatial/cuspatial/core/geodataframe.py:187]: Note that cudf's [`_apply_boolean_mask`](https://github.com/rapidsai/cudf/blob/45763fa1f90c174ff9123b4839c729e6c0dee700/python/cudf/cudf/core/indexed_frame.py#L3017) now also accepts a `keep_index` keyword argument

Comment (isVoid) [reply python/cuspatial/cuspatial/core/geodataframe.py]: cuspatial should really just try to avoid depending on cudf internals...

**Review by wence- — APPROVED** [review]

File: python/cuspatial/cuspatial/core/geodataframe.py (hunk 1 above)
Comment (wence-) [reply python/cuspatial/cuspatial/core/geodataframe.py]: I guess that is difficult if you're subclassing though :(

**Review by harrism — APPROVED** [review]

**Review by thomcom — APPROVED** [review]

**Pull request merged**